sw/device/lib/dif/dif_pinmux.c

```diff
@@ -0,0 +1,232 @@
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_INSEL_RESET_DEFAULT);
+  }
+
+  // Reset all of the outsel registers (multireg) to the reset value.
+  for (int i = 0; i < pinmux->priv->outsel_register_num; ++i) {
+    ptrdiff_t register_offset =
+        PINMUX_MIO_OUTSEL0_REG_OFFSET + (i * PINMUX_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(pinmux->base_addr, register_offset,
+                        PINMUX_OUTSEL_RESET_DEFAULT);
+  }
+}
+
+dif_pinmux_init_error_t dif_pinmux_init(mmio_region_t base_addr,
+                                        dif_pinmux_t *pinmux) {
+  if (pinmux == NULL) {
+    return kDifPinmuxInitErrorInvalidArgument;
+  }
+
+  // Check whether the peipheral can be configured.
+  bool write_enabled = mmio_region_get_bit32(base_addr, PINMUX_REGEN_REG_OFFSET,
+                                             PINMUX_REGEN_WEN);
+  if (!write_enabled) {
+    return kDifPinmuxInitErrorRegistersLocked;
+  }
+
+  pinmux->base_addr = base_addr;
+  pinmux->write_enabled = write_enabled;
+
+  // Calculate HW information once on initialisation.
+  pinmux_private.insel_field_width =
+      pinmux_bits_per_value(DIF_PINMUX_INSEL_NUM);
+
+  pinmux_private.outsel_field_width =
+      pinmux_bits_per_value(DIF_PINMUX_OUTSEL_NUM);
+
+  pinmux_private.insel_fields_per_register =
+      PINMUX_REGISTER_WIDTH / pinmux_private.insel_field_width;
+
+  pinmux_private.outsel_fields_per_register =
+      PINMUX_REGISTER_WIDTH / pinmux_private.outsel_field_width;
+
+  pinmux_private.insel_register_num = pinmux_register_num(
+      DIF_PINMUX_PERIPHERAL_IN_NUM, pinmux_private.insel_fields_per_register);
+
+  pinmux_private.outsel_register_num = pinmux_register_num(
+      DIF_PINMUX_MIO_IN_NUM, pinmux_private.outsel_fields_per_register);
+
+  pinmux->priv = &pinmux_private;
+
+  // Set the peripheral to the reset defaults.
+  dif_pinmux_reset(pinmux);
+
+  return kDifPinmuxInitErrorNone;
+}
+
+dif_pinmux_registers_lock_error_t dif_pinmux_registers_lock(
+    dif_pinmux_t *pinmux) {
+  if (pinmux == NULL) {
+    return kDifPinmuxRegistersLockErrorInvalidArgument;
+  }
+
+  mmio_region_write32(pinmux->base_addr, PINMUX_REGEN_REG_OFFSET, 0);
+  pinmux->write_enabled = false;
+
+  return kDifPinmuxRegistersLockErrorNone;
+}
+
+dif_pinmux_insel_set_error_t dif_pinmux_insel_set(
+    const dif_pinmux_t *pinmux, dif_pinmux_peripheral_in_t peripheral_input,
+    dif_pinmux_insel_t mio_select) {
+  if (pinmux == NULL) {
+    return kDifPinmuxInselSetErrorInvalidArgument;
+  }
+
+  if (!pinmux->write_enabled) {
+    return kDifPinmuxInselSetErrorRegistersLocked;
+  }
+
+  if (mio_select < kDifPinmuxInselTieToZero ||
+      mio_select > kDifPinmuxInselPadLast) {
+    return kDifPinmuxInselSetErrorBadMioSelect;
+  }
+
+  // Register offset based on a given peripheral input and MIO select.
+  size_t register_index =
+      (size_t)peripheral_input / pinmux->priv->insel_fields_per_register;
```

Comment:
I'd add some parentheses here to make the desired scope of the cast clear. I can never remember the correct precedence rules.

Reply:
Done.